### ISSUE TITLE
More checkers

### ISSUE THREAD

**Issue by stefanitsky** [DudeNr33/gherlint#16]
Hi, good work!

Can we add more checkers?

I used [this](https://github.com/vsiakka/gherkin-lint) npm package, but it's not comfortable to use npm in python repository, I want to improve this package to use it instead of npm package.

Do I understand correctly, that issue codes (like W101, W102 etc.) are custom codes? Can I add my own codes and checkers? What are the naming rules?

**Comment by DudeNr33** [comment]
Thank you!

Yes, the codes are custom. 
`gherlint` currently knows about four categories: 
* E: error
* W: warning
* C: convention
* R: refactor

Message IDs (e.g. W101) and message names (e.g. `missing-feature-name`) must be unique. Ideally all messages from one checker should start with the same first digit, and the last two digits are just running numbers starting at 1.
However this is not enforced at the moment.

I will happily review and accept PRs. 👍  
You can either extend the existing checkers if suitable, or create a new module under `gherlint/checkers`. You only need to add a `register_checker` function in this file which returns the instantiated checker - just look at one of the existing checkers as an example. 

For testing, functional tests are the easiest way to verify the checker works as intended. You can put a new feature file under `tests/functional/<first_letter_of_message_name>/<message_name>.feature`, together with a `<message_name>.txt` file containing the expected output.
The test suite should find the new test file automatically.

Let me know if you need more help, or have ideas for new checks that you'd be interested in but don't know how to implement or don't have the time for. I will see what I can do!

**Comment by DudeNr33** [comment]
I think `required-tag-missing` and `use-of-restricted-pattern` sound reasonable.

There are pros and cons for having one class per check.
The downside surely is that reusing existing code (and state information a checker collects during runtime) for multiple checks will be harder. It would also be harder if you have checks that are very closely related and would require you to do the same intermediary steps twice if they are located in different checkers. Also the object creation could have performance impacts when the number of checks grows.

Are there other benefits you see besides not having to think where to put it? (For the examples above, I would create a `ConventionsChecker`, as both checks are related to user-defined conventions). 
It would certainly make sense to keep a clear focus for each checker (also regarding single responsibility principle), but I don't think it has to be a convention that there must only be one check per checker class.

Beside that, the more pressing problem in context with those two checks is that `gherlint` currently does not have support for a configuration file, which would be necessary for these checks.

**Comment by stefanitsky** [comment]
I agree, that's what I wanted, configuration with `pyproject.toml` file.

**Comment by DudeNr33** [comment]
The draft is at #17, feel free to give feedback if this also suits your needs.

**Comment by stefanitsky** [comment]
My turn 😋

**Comment by stefanitsky** [comment]
Hm, I faced with some problem:
I want to write functional test, but moreover, I need a pyproject.toml, like .feature and .txt files.

What do you think, how can we implement this?

**Comment by DudeNr33** [comment]
We have to extend `test_functional.py`.
It should be possible to simply check if a config file for this test exists, then request the config once in `test_expected_outcome` before the `GherkinLinter` is instantiated, and then reset it at the end of the test.

I can check if that works and prepare a PR tomorrow.

**Comment by stefanitsky** [comment]
I tried to implement something like you said, but I faced with some trouble, that I don't know which is the best solution to load config (mock, fake, or modify GherkinLinter class).

The easiest solution that I found: override `_init_options` method and add custom arg `config` to `__init__` method. What do you think?

**Comment by DudeNr33** [comment]
My idea was to just use the caching mechanism in `Config` to make it work.
We can simply define a contextmanager `use_config` which requests the config before the functional test once and therefore sets the cache. After the test is finished, the cached config is reset.

```python
# pylint: disable=protected-access
@contextmanager
def use_config(config: Path = None):
    """Populates the cached config instance with the config specified for this test."""
    Config._config = None
    Config.get_config(config)
    try:
        yield
    finally:
        # reset cached config
        Config._config = None


@pytest.mark.parametrize(
    "feature_file, expected_output", list(zip(feature_files, result_files)), ids=idfn
)
def test_expected_outcome(feature_file: Path, expected_output: Path, capsys) -> None:
    if feature_file.with_suffix(".toml").exists():
        config = feature_file.with_suffix(".toml")
    else:
        config = None
    with use_config(config):
        linter = GherkinLinter(feature_file)
        linter.run()
    captured = capsys.readouterr()
    output = _patch_filename(feature_file, captured.out)
    assert output.strip() == expected_output.read_text("utf8").strip()
```

**Comment by stefanitsky** [comment]
Ah, right, I forgot or didn't see, that `Config` class uses cache.

Thanks, I will try this approach 👍

**Comment by stefanitsky** [comment]
What do you think abount [Pydantic](https://github.com/samuelcolvin/pydantic/)? Can we replace `Options` class with Pydantic model? I need more complex parse, than simple types. For example, if I want to parse list of strings, I just can't, because I need to extend `Options` class and parse it manually.

Now:
```py
class Options:
    """Options for a checker."""

    config_section: str = "default"

    def __init__(self, **kwargs):
        for name, value in kwargs.items():
            annotations = getattr(self, "__annotations__", {})
            if name in annotations:
                expected_type = annotations[name]
                if not isinstance(value, expected_type):
                    raise ValueError(
                        f"Invalid type for option '{name}': "
                        f"expected '{expected_type}', got '{type(value)}'"
                    )
                setattr(self, name, value)

    @classmethod
    def from_config(cls) -> Options:
        """Create options from the config."""
        config = Config.get_config()
        options = config.get(cls.config_section, {})
        return cls(**options)

```

With Pydantic:
```py
class Options(BaseModel):
    """Options for a checker."""

    config_section: ClassVar[str] = Field(..., description="config section with options for a checker")

    @classmethod
    def from_config(cls) -> Options:
        """Create options from the config."""
        config = Config.get_config()
        options = config.get(cls.config_section, {})
        return cls(**options)
```

So, I can write complex options classes, for example:
```py
class ProhibitedNamingPatternsOptions(BaseModel):
    feature: Optional[str]
    background: Optional[str]
    scenario: Optional[str]


class ConventionCheckerOptions(Options):
    config_section = "convention"

    required_feature_tags: List[str] = []
    prohibited_naming_patterns: Optional[ProhibitedNamingPatternsOptions]
```

TOML example:
```toml
[convention]
required_feature_tags = ["required_1"]
prohibited_naming_patterns = {feature = "request|databas[e|es]|db|*test*"}

# or
# [convention.prohibited_naming_patterns]
# feature = "request|databas[e|es]|db|*test*"
```

**Comment by DudeNr33** [comment]
I see that the simple `isinstance` check of the current implementation is causing trouble for complex types. 
I have no prior experience with `pydantic`, but it looks like it would work well.

Would you like to create a separate PR for this first? That will make it easier to review. It could already include the change for using options in functional tests as outlined above.

**Comment by stefanitsky** [comment]
Simple `isinstance` check is not working due to exception `TypeError: Subscripted generics cannot be used with class and instance checks`. That's why I used Pydantic.

I can create a separate PR for this, but I need a simple solution to fix this, any ideas? Unfortunately I didn't come up with anything 😢

**Comment by DudeNr33** [comment]
Maybe there is a misunderstanding.
I am fine with switching to `pydantic`. Your `With pydantic` example above looks good to me, and `pydantic` itself does not have a lot of dependencies itself, so including it is no problem.
The troublesome `isinstance` check would no longer be necessary, as `pydantic` takes care of the validation.

My suggestion is that you first create a separate PR that implements the `Options` with `pydantic`, and then continue on the work of the actual `ConventionsChecker`.

Let me know if I can help you with that, or if you meant something else.
